### PR TITLE
Implement reset password

### LINKED ISSUE #39: Login - Forgot Password

### PR DESCRIPTION
**Pull request by DeKleineKobini**
Closes #39

### CODE REVIEW

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/AuthenticationController.java [hunk 1]
```diff
@@ -67,4 +68,21 @@ public MessageResponse signUp(@Valid @RequestBody SignupRequest signUpRequest) {
 
         return new MessageResponse("User registered successfully!");
     }
+
+    @PostMapping("/forgotPassword")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void forgotPassword(@Valid @RequestBody ForgotRequest forgotRequest) {
```
Comment (AhmadAlhajKarim) [inline inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/AuthenticationController.java:74]: Is a dto really necessary here?

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/entities/ResetCode.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        this.account = account;
+    }
+
+    public boolean isValid(LocalDateTime on) {
+        return on.isBefore(expiryDate);
+    }
+
+    @PrePersist
+    protected void onCreate() {
+        setExpiryDate(LocalDateTime.now().plusDays(1));
```
Comment (AhmadAlhajKarim) [inline inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/entities/ResetCode.java:53]: Mostly reset password link/code is valid for few hours but this also work

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/services/MailService.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    public void sendResetPasswordLink(String to, String code) throws MessagingException {
+        MimeMessage message = mailSender.createMimeMessage();
+
+        MimeMessageHelper helper = new MimeMessageHelper(message, true);
+
+        // FIXME - Improve email from.
+        helper.setFrom("noreply@realjobs.com");
+        helper.setTo(to);
+        helper.setSubject("Request reset password");
+        // FIXME - Improve email content.
```
Comment (TomD011099) [inline inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/services/MailService.java]: Plz fix

File: inetum-realjobs-backend/src/test/resources/application.properties [hunk 4]
```diff
@@ -3,3 +3,8 @@ spring.datasource.username=test
 spring.datasource.password=test
 spring.jpa.hibernate.ddl-auto=create-drop
 spring.sql.init.mode=never
+spring.mail.host=localhost
+spring.mail.port=3025
+spring.mail.username=mail
+spring.mail.password=test
+spring.mail.test-connection=true
```
Comment (AhmadAlhajKarim) [inline inetum-realjobs-backend/src/test/resources/application.properties]: Does this meant to be pushed on git?

File: inetum-realjobs-backend/src/test/java/world/inetum/realdolmen/realjobs/controllers/AuthenticationControllerIT.java [hunk 5]
```diff
@@ -197,4 +199,80 @@ void registerWithFakeCountry() {
...
+                        .content(asJsonString(forgotRequest)))
+                .andExpect(status().isNoContent());
+
+        greenMail.waitForIncomingEmail(1);
+
+        var allCodes = resetCodeRepository.findAll();
+        assertEquals(1, allCodes.size(), "There should be a single reset code in the database");
+
+        var code = allCodes.get(0);
+        assertEquals("test@realjobs.com", code.getAccount().getEmail(), "Code should be linked to the given account");
```
Comment (TomD011099) [inline inetum-realjobs-backend/src/test/java/world/inetum/realdolmen/realjobs/controllers/AuthenticationControllerIT.java:221]: Maybe also check if UUIDs are the same?

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/entities/ResetCode.java (hunk 2 above)
Comment (DeKleineKobini) [reply inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/entities/ResetCode.java:53]: The analysis specified 24h.

File: inetum-realjobs-backend/src/test/resources/application.properties (hunk 4 above)
Comment (DeKleineKobini) [reply inetum-realjobs-backend/src/test/resources/application.properties]: Yes, these are for GreenMail, which runs alongside the tests.

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/AuthenticationController.java (hunk 1 above)
Comment (DeKleineKobini) [reply inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/AuthenticationController.java:74]: Yes, the only other way would be to directly send the email as the body, but that's a fairly bad practice. Sending it as JSON (and thus using a dto) is the best imo.

**Review by TomD011099 — CHANGES_REQUESTED** [review]

**Review by AhmadAlhajKarim — APPROVED** [review]

**Review by TomD011099 — APPROVED** [review]

**Pull request merged**